src/main/java/de/tum/in/www1/artemis/service/user/UserService.java

```diff
@@ -236,10 +240,26 @@ public User registerUser(UserDTO userDTO, String password) {
         Set<Authority> authorities = new HashSet<>();
         authorityRepository.findById(STUDENT.getAuthority()).ifPresent(authorities::add);
         newUser.setAuthorities(authorities);
-        saveUser(newUser);
+
         // we need to save first so that the user can be found in the database in the subsequent method
-        optionalVcsUserManagementService.ifPresent(vcsUserManagementService -> vcsUserManagementService.createVcsUser(newUser));
-        optionalCIUserManagementService.ifPresent(ciUserManagementService -> ciUserManagementService.createUser(newUser));
+        saveUser(newUser);
+
+        optionalVcsUserManagementService.ifPresent(vcsUserManagementService -> {
+            if (!vcsUserManagementService.canCreateVcsUser(newUser)) {
+                // Delete the user that was previously saved and throw
+                deleteUser(newUser);
+                throw new VCSAccountExistsException();
+            }
+        });
+
+        optionalCIUserManagementService.ifPresent(ciUserManagementService -> {
+            if (!ciUserManagementService.canCreateUser(newUser)) {
+                // Delete the user that was previously saved and throw
+                deleteUser(newUser);
```

Comment:
are you sure there can be no edge cases here where you actually would delete an actual user here?
What happens if an administrator creates an account manually between the registration and activation of a user account?